src/ui/legendWidget.js

```diff
@@ -0,0 +1,261 @@
+
+    m_categories.forEach(function (d) {
+      test.text(d.name);
+      width = Math.max(width, test.node().getBBox().width);
+    });
+    test.remove();
+
+    height = m_spacing * (m_categories.length + 1);
+    return {
+      width: width + 50,
+      height: height
+    };
+  };
+
+  //////////////////////////////////////////////////////////////////////////////
+  /**
+   * Redraw the legend
+   */
+  //////////////////////////////////////////////////////////////////////////////
+  this.draw = function () {
+
+    m_this._init();
+    function applyColor(selection) {
+      selection.style('fill', function (d) {
+          if (d.style.fill || d.style.fill === undefined) {
+            return d.style.fillColor;
+          } else {
+            return 'none';
+          }
+        })
+        .style('fill-opacity', function (d) {
+          if (d.style.fillOpacity === undefined) {
+            return 1;
+          }
+          return d.style.fillOpacity;
+        })
+        .style('stroke', function (d) {
+          if (d.style.stroke || d.style.stroke === undefined) {
+            return d.style.strokeColor;
+          } else {
+            return 'none';
+          }
+        })
+        .style('stroke-opacity', function (d) {
+          if (d.style.strokeOpacity === undefined) {
+            return 1;
+          }
+          return d.style.strokeOpacity;
+        })
+        .style('stroke-width', function (d) {
+          if (d.style.strokeWidth === undefined) {
+            return 1.5;
+          }
+          return d.style.strokeWidth;
+        });
+    }
+
+    m_border.attr('height', m_this.size().height + 2 * m_padding)
+      .style('display', null);
+
+    var scale = m_this._scale();
+
+    var labels = m_group.selectAll('g.geo-label')
+      .data(m_categories, function (d) { return d.name; });
+
+    var g = labels.enter().append('g')
+      .attr('class', 'geo-label')
+      .attr('transform', function (d, i) {
+        return 'translate(0,' + scale.y(i) + ')';
+      });
+
+    applyColor(g.filter(function (d) {
+        return d.type !== 'point' && d.type !== 'line';
+      }).append('rect')
+        .attr('x', 0)
+        .attr('y', -6)
+        .attr('rx', 5)
+        .attr('ry', 5)
+        .attr('width', 40)
+        .attr('height', 12)
+    );
+
+    applyColor(g.filter(function (d) {
+        return d.type === 'point';
+      }).append('circle')
+        .attr('cx', 20)
+        .attr('cy', 0)
+        .attr('r', 6)
+    );
+
+    applyColor(g.filter(function (d) {
+        return d.type === 'line';
+      }).append('line')
+        .attr('x1', 0)
+        .attr('y1', 0)
+        .attr('x2', 40)
+        .attr('y2', 0)
+    );
+
+    g.append('text')
+      .attr('x', '50px')
+      .attr('y', 0)
+      .attr('dy', '0.3em')
+      .text(function (d) {
+        return d.name;
+      });
+
+    return m_this;
+  };
+
+  //////////////////////////////////////////////////////////////////////////////
+  /**
```

Comment:
this is awesome. ..at some point we have to get this information from a look up table (once we get it from the cinema). Also, we may have to support gradients but all that could be done later (or may be in a derived class).